File: packages/android/src/main/java/com/kozmos/components/NavigationItem/NavigationItem.kt
package com.kozmos.components.navigationitem

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.defaultMinSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.LocalContentColor
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.disabled
import androidx.compose.ui.semantics.selected
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import com.kozmos.tokens.KozmosDimensions
import com.kozmos.tokens.KozmosThemeTokens

enum class KozmosNavigationItemPlacement {
    Top,
    Side,
    Rail
}

enum class KozmosNavigationItemDensity {
    Default,
    Compact
}

enum class KozmosNavigationItemContent {
    Label,
    IconLabel,
    IconOnly,
    Badge,
    Trailing
}

enum class KozmosNavigationItemState {
    Default,
    Hover,
    Selected,
    Focus,
    Disabled
}

@Composable
fun KozmosNavigationItem(
    label: String? = null,
    modifier: Modifier = Modifier,
    placement: KozmosNavigationItemPlacement = KozmosNavigationItemPlacement.Side,
    density: KozmosNavigationItemDensity = KozmosNavigationItemDensity.Default,
    content: KozmosNavigationItemContent = KozmosNavigationItemContent.Label,
    state: KozmosNavigationItemState = KozmosNavigationItemState.Default,
    selected: Boolean = false,
    enabled: Boolean = true,
    focusVisible: Boolean = false,
    onClick: () -> Unit = {},
    icon: (@Composable () -> Unit)? = null,
    badge: (@Composable () -> Unit)? = null,
    trailing: (@Composable () -> Unit)? = null
) {
    val isSelected = selected || state == KozmosNavigationItemState.Selected
    val isDisabled = !enabled || state == KozmosNavigationItemState.Disabled
    val isFocusVisible = focusVisible || state == KozmosNavigationItemState.Focus
    val shape = RoundedCornerShape(KozmosDimensions.primitivesLayoutRadius100)
    val minHeight = when (placement) {
        KozmosNavigationItemPlacement.Rail -> if (density == KozmosNavigationItemDensity.Compact) 64.dp else 72.dp
        else -> 44.dp
    }
    val sizeModifier = when (placement) {
        KozmosNavigationItemPlacement.Side -> Modifier
            .fillMaxWidth()
            .defaultMinSize(minHeight = minHeight)
        KozmosNavigationItemPlacement.Rail -> Modifier
            .width(if (density == KozmosNavigationItemDensity.Compact) 64.dp else 72.dp)
            .defaultMinSize(minHeight = minHeight)
        KozmosNavigationItemPlacement.Top -> Modifier.defaultMinSize(minHeight = minHeight)
    }
    val backgroundColor = when {
        isSelected || state == KozmosNavigationItemState.Hover || state == KozmosNavigationItemState.Focus ->
            KozmosThemeTokens.primitivesColorsBackground100
        else -> Color.Transparent
    }
    val contentColor = when {
        isDisabled -> KozmosThemeTokens.primitivesColorsForeground500
        isSelected -> KozmosThemeTokens.primitivesColorsTheme500
        else -> KozmosThemeTokens.primitivesColorsForeground100
    }
    val horizontalPadding = when {
        placement == KozmosNavigationItemPlacement.Rail -> 8.dp
        density == KozmosNavigationItemDensity.Compact -> 10.dp
        else -> 12.dp
    }
    val verticalPadding = if (placement == KozmosNavigationItemPlacement.Rail) 8.dp else 8.dp

    val rootModifier = modifier
        .then(sizeModifier)
        .clip(shape)
        .background(backgroundColor, shape)
        .then(
            if (isFocusVisible) {
                Modifier.border(2.dp, KozmosThemeTokens.primitivesColorsTheme500, shape)
            } else {
                Modifier
            }
        )
        .clickable(
            enabled = !isDisabled,
            role = Role.Button,
            onClick = onClick
        )
        .semantics {
            if (isSelected) this.selected = true
            if (isDisabled) disabled()
        }
        .padding(horizontal = horizontalPadding, vertical = verticalPadding)

    CompositionLocalProvider(LocalContentColor provides contentColor) {
        if (placement == KozmosNavigationItemPlacement.Rail) {
            RailNavigationItemContent(
                label = label,
                content = content,
                icon = icon,
                modifier = rootModifier
            )
        } else {
            RowNavigationItemContent(
                label = label,
                placement = placement,
                content = content,
                icon = icon,
                badge = badge,
                trailing = trailing,
                modifier = rootModifier
            )
        }
    }
}

@Composable
private fun RowNavigationItemContent(
    label: String?,
    placement: KozmosNavigationItemPlacement,
    content: KozmosNavigationItemContent,
    icon: (@Composable () -> Unit)?,
    badge: (@Composable () -> Unit)?,
    trailing: (@Composable () -> Unit)?,
    modifier: Modifier = Modifier
) {
    Row(
        modifier = modifier,
        horizontalArrangement = Arrangement.spacedBy(KozmosDimensions.primitivesLayoutSpacing100),
        verticalAlignment = Alignment.CenterVertically
    ) {
        if (shouldRenderIcon(content) && icon != null) {
            Box(
                modifier = Modifier.size(20.dp),
                contentAlignment = Alignment.Center
            ) {
                icon()
            }
        }

        if (shouldRenderLabel(content) && label != null) {
            Text(
                text = label,
                modifier = if (placement == KozmosNavigationItemPlacement.Side) Modifier.weight(1f) else Modifier,
                color = LocalContentColor.current,
                maxLines = 1,
                overflow = TextOverflow.Ellipsis,
                style = MaterialTheme.typography.labelLarge,
                fontWeight = FontWeight.Medium
            )
        }

        if (content == KozmosNavigationItemContent.Badge && badge != null) {
            Box(
                modifier = Modifier
                    .defaultMinSize(minHeight = 20.dp)
                    .clip(RoundedCornerShape(999.dp))
                    .background(KozmosThemeTokens.primitivesColorsBackground0)
                    .border(1.dp, KozmosThemeTokens.primitivesColorsBackground200, RoundedCornerShape(999.dp))
                    .padding(horizontal = 6.dp),
                contentAlignment = Alignment.Center
            ) {
                CompositionLocalProvider(LocalContentColor provides KozmosThemeTokens.primitivesColorsForeground100) {
                    badge()
                }
            }
        }

        if (content == KozmosNavigationItemContent.Trailing && trailing != null) {
            CompositionLocalProvider(LocalContentColor provides KozmosThemeTokens.primitivesColorsForeground500) {
                trailing()
            }
        }
    }
}

@Composable
private fun RailNavigationItemContent(
    label: String?,
    content: KozmosNavigationItemContent,
    icon: (@Composable () -> Unit)?,
    modifier: Modifier = Modifier
) {
    Column(
        modifier = modifier,
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(KozmosDimensions.primitivesLayoutSpacing50)
    ) {
        if (shouldRenderIcon(content) && icon != null) {
            Box(
                modifier = Modifier.size(24.dp),
                contentAlignment = Alignment.Center
            ) {
                icon()
            }
        }

        if (shouldRenderLabel(content) && label != null) {
            Text(
                text = label,
                color = LocalContentColor.current,
                maxLines = 1,
                overflow = TextOverflow.Ellipsis,
                textAlign = TextAlign.Center,
                style = MaterialTheme.typography.labelSmall,
                fontWeight = FontWeight.SemiBold
            )
        }
    }
}

private fun shouldRenderIcon(content: KozmosNavigationItemContent): Boolean {
    return content != KozmosNavigationItemContent.Label
}

private fun shouldRenderLabel(content: KozmosNavigationItemContent): Boolean {
    return content != KozmosNavigationItemContent.IconOnly
}

File: packages/ios/Sources/Components/NavigationItem/NavigationItem.swift
import SwiftUI

public enum KozmosNavigationItemPlacement {
    case top
    case side
    case rail
}

public enum KozmosNavigationItemDensity {
    case `default`
    case compact
}

public enum KozmosNavigationItemContent {
    case label
    case iconLabel
    case iconOnly
    case badge
    case trailing
}

public enum KozmosNavigationItemState {
    case `default`
    case hover
    case selected
    case focus
    case disabled
}

public struct KozmosNavigationItem: View {
    let label: String?
    let placement: KozmosNavigationItemPlacement
    let density: KozmosNavigationItemDensity
    let content: KozmosNavigationItemContent
    let state: KozmosNavigationItemState
    let selected: Bool
    let disabled: Bool
    let focusVisible: Bool
    let action: () -> Void
    private let icon: AnyView?
    private let badge: AnyView?
    private let trailing: AnyView?

    public init(
        label: String? = nil,
        placement: KozmosNavigationItemPlacement = .side,
        density: KozmosNavigationItemDensity = .default,
        content: KozmosNavigationItemContent = .label,
        state: KozmosNavigationItemState = .default,
        selected: Bool = false,
        disabled: Bool = false,
        focusVisible: Bool = false,
        action: @escaping () -> Void = {}
    ) {
        self.init(
            label: label,
            placement: placement,
            density: density,
            content: content,
            state: state,
            selected: selected,
            disabled: disabled,
            focusVisible: focusVisible,
            icon: nil,
            badge: nil,
            trailing: nil,
            action: action
        )
    }

    public init<Icon: View>(
        label: String? = nil,
        placement: KozmosNavigationItemPlacement = .side,
        density: KozmosNavigationItemDensity = .default,
        content: KozmosNavigationItemContent = .iconLabel,
        state: KozmosNavigationItemState = .default,
        selected: Bool = false,
        disabled: Bool = false,
        focusVisible: Bool = false,
        action: @escaping () -> Void = {},
        @ViewBuilder icon: () -> Icon
    ) {
        self.init(
            label: label,
            placement: placement,
            density: density,
            content: content,
            state: state,
            selected: selected,
            disabled: disabled,
            focusVisible: focusVisible,
            icon: AnyView(icon()),
            badge: nil,
            trailing: nil,
            action: action
        )
    }

    public init<Icon: View, Badge: View>(
        label: String? = nil,
        placement: KozmosNavigationItemPlacement = .side,
        density: KozmosNavigationItemDensity = .default,
        state: KozmosNavigationItemState = .default,
        selected: Bool = false,
        disabled: Bool = false,
        focusVisible: Bool = false,
        action: @escaping () -> Void = {},
        @ViewBuilder icon: () -> Icon,
        @ViewBuilder badge: () -> Badge
    ) {
        self.init(
            label: label,
            placement: placement,
            density: density,
            content: .badge,
            state: state,
            selected: selected,
            disabled: disabled,
            focusVisible: focusVisible,
            icon: AnyView(icon()),
            badge: AnyView(badge()),
            trailing: nil,
            action: action
        )
    }

    public init<Icon: View, Trailing: View>(
        label: String? = nil,
        placement: KozmosNavigationItemPlacement = .side,
        density: KozmosNavigationItemDensity = .default,
        state: KozmosNavigationItemState = .default,
        selected: Bool = false,
        disabled: Bool = false,
        focusVisible: Bool = false,
        action: @escaping () -> Void = {},
        @ViewBuilder icon: () -> Icon,
        @ViewBuilder trailing: () -> Trailing
    ) {
        self.init(
            label: label,
            placement: placement,
            density: density,
            content: .trailing,
            state: state,
            selected: selected,
            disabled: disabled,
            focusVisible: focusVisible,
            icon: AnyView(icon()),
            badge: nil,
            trailing: AnyView(trailing()),
            action: action
        )
    }

    private init(
        label: String?,
        placement: KozmosNavigationItemPlacement,
        density: KozmosNavigationItemDensity,
        content: KozmosNavigationItemContent,
        state: KozmosNavigationItemState,
        selected: Bool,
        disabled: Bool,
        focusVisible: Bool,
        icon: AnyView?,
        badge: AnyView?,
        trailing: AnyView?,
        action: @escaping () -> Void
    ) {
        self.label = label
        self.placement = placement
        self.density = density
        self.content = content
        self.state = state
        self.selected = selected
        self.disabled = disabled
        self.focusVisible = focusVisible
        self.icon = icon
        self.badge = badge
        self.trailing = trailing
        self.action = action
    }

    public var body: some View {
        Button(action: action) {
            Group {
                if placement == .rail {
                    railContent
                } else {
                    rowContent
                }
            }
            .padding(contentPadding)
            .frame(width: fixedWidth)
            .frame(minHeight: minHeight)
            .background(backgroundColor)
            .foregroundColor(foregroundColor)
            .clipShape(shape)
            .overlay(
                shape.stroke(focusRingColor, lineWidth: isFocusVisible ? 2 : 0)
            )
        }
        .buttonStyle(.plain)
        .disabled(isDisabled)
        .opacity(isDisabled ? 0.5 : 1)
        .accessibilityLabel(label ?? "")
        .accessibilityValue(isSelected ? "Selected" : "")
    }

    private var rowContent: some View {
        HStack(spacing: KozmosDimensions.primitivesLayoutSpacing100) {
            if shouldRenderIcon, let icon = icon {
                icon
                    .frame(width: 20, height: 20)
            }

            if shouldRenderLabel, let label = label {
                Text(label)
                    .font(.subheadline)
                    .fontWeight(.medium)
                    .lineLimit(1)
                    .truncationMode(.tail)
                    .frame(
                        maxWidth: placement == .side ? .infinity : nil,
                        alignment: .leading
                    )
            }

            if shouldRenderBadge, let badge = badge {
                badge
                    .font(.caption)
                    .padding(.horizontal, 6)
                    .frame(minHeight: 20)
                    .background(KozmosColors.primitivesColorsBackground0)
                    .clipShape(Capsule())
                    .overlay(Capsule().stroke(KozmosColors.primitivesColorsBackground200, lineWidth: 1))
            }

            if shouldRenderTrailing, let trailing = trailing {
                trailing
                    .foregroundColor(KozmosColors.primitivesColorsForeground500)
            }
        }
    }

    private var railContent: some View {
        VStack(spacing: KozmosDimensions.primitivesLayoutSpacing50) {
            if shouldRenderIcon, let icon = icon {
                icon
                    .frame(width: 24, height: 24)
            }

            if shouldRenderLabel, let label = label {
                Text(label)
                    .font(.caption)
                    .fontWeight(.semibold)
                    .lineLimit(1)
                    .truncationMode(.tail)
                    .multilineTextAlignment(.center)
            }
        }
        .frame(maxWidth: .infinity)
    }

    private var isSelected: Bool {
        selected || state == .selected
    }

    private var isDisabled: Bool {
        disabled || state == .disabled
    }

    private var isFocusVisible: Bool {
        focusVisible || state == .focus
    }

    private var shouldRenderIcon: Bool {
        content != .label && icon != nil
    }

    private var shouldRenderLabel: Bool {
        content != .iconOnly
    }

    private var shouldRenderBadge: Bool {
        content == .badge && badge != nil
    }

    private var shouldRenderTrailing: Bool {
        content == .trailing && trailing != nil
    }

    private var minHeight: CGFloat {
        if placement == .rail {
            return density == .compact ? 64 : 72
        }
        return 44
    }

    private var fixedWidth: CGFloat? {
        switch placement {
        case .rail:
            return density == .compact ? 64 : 72
        case .side:
            return nil
        case .top:
            return nil
        }
    }

    private var contentPadding: EdgeInsets {
        if placement == .rail {
            return EdgeInsets(top: 8, leading: 8, bottom: 8, trailing: 8)
        }

        let horizontal = density == .compact ? 10.0 : 12.0
        return EdgeInsets(top: 8, leading: horizontal, bottom: 8, trailing: horizontal)
    }

    private var shape: RoundedRectangle {
        RoundedRectangle(cornerRadius: KozmosDimensions.primitivesLayoutRadius100)
    }

    private var backgroundColor: Color {
        if isSelected || state == .hover || state == .focus {
            return KozmosColors.primitivesColorsBackground100
        }
        return Color.clear
    }

    private var foregroundColor: Color {
        if isDisabled {
            return KozmosColors.primitivesColorsForeground500
        }
        if isSelected {
            return KozmosColors.primitivesColorsTheme500
        }
        return KozmosColors.primitivesColorsForeground100
    }

    private var focusRingColor: Color {
        isFocusVisible ? KozmosColors.primitivesColorsTheme500 : Color.clear
    }
}
